pkg/iac/rego/load.go

```diff
@@ -251,50 +257,102 @@ func (s *Scanner) compilePolicies(srcFS fs.FS, paths []string) error {
+		s.logger.Warn(
+			"Module has legacy metadata format - please update to use annotations",
+			log.FilePath(module.Package.Location.File),
+			log.String("details", doc.URL("/docs/scanner/misconfiguration/custom", "metadata")),
+		)
+		return
+	}
+
+	metadata, err := MetadataFromAnnotations(module)
+	if err != nil {
+		s.logger.Error(
+			"Failed to retrieve metadata from annotations",
+			log.FilePath(module.Package.Location.File),
+			log.Err(err),
+		)
+		return
+	}
 
+	if metadata != nil {
+		s.moduleMetadata[path] = metadata
+	}
+}
+
+// moduleHasLegacyMetadataFormat checks if the module has an legacy metadata format.
+// Returns true if the metadata is represented as a “__rego_metadata__” rule,
+// which was used before annotations were introduced.
+func moduleHasLegacyMetadataFormat(module *ast.Module) bool {
+	return slices.ContainsFunc(module.Rules, func(rule *ast.Rule) bool {
+		return rule.Head.Name.Equal(ast.Var("__rego_metadata__"))
+	})
+}
+
+// moduleHasLegacyInputFormat checks if the module has an legacy input format.
+// Returns true if the input is represented as a “__rego_input__” rule,
+// which was used before annotations were introduced.
+func moduleHasLegacyInputFormat(module *ast.Module) bool {
+	return slices.ContainsFunc(module.Rules, func(rule *ast.Rule) bool {
+		return rule.Head.Name.Equal(ast.Var("__rego_input__"))
+	})
+}
+
+// filterModules filters the Rego modules based on metadata.
+func (s *Scanner) filterModules() error {
 	filtered := make(map[string]*ast.Module)
 	for name, module := range s.policies {
-		meta, err := retriever.RetrieveMetadata(context.TODO(), module)
+		metadata, err := s.metadataForModule(context.Background(), name, module, nil)
 		if err != nil {
-			return err
+			return fmt.Errorf("retrieve metadata for module %s: %w", name, err)
 		}
 
-		if !meta.hasAnyFramework(s.frameworks) {
-			continue
-		}
-
-		if IsBuiltinNamespace(getModuleNamespace(module)) {
-			if s.disabledCheckIDs.Contains(meta.ID) { // ignore builtin disabled checks
-				continue
-			}
-		}
-
-		if len(meta.InputOptions.Selectors) == 0 {
-			if !meta.Library {
-				s.logger.Warn(
-					"Module has no input selectors - it will be loaded for all inputs!",
-					log.FilePath(module.Package.Location.File),
-					log.String("module", name),
-				)
-			}
+		if s.isModuleApplicable(module, metadata, name) {
 			filtered[name] = module
-			continue
 		}
-
-		filtered[name] = module
 	}
 
 	s.policies = filtered
 	return nil
 }
+
+func (s *Scanner) isModuleApplicable(module *ast.Module, metadata *StaticMetadata, name string) bool {
+	if !metadata.hasAnyFramework(s.frameworks) {
+		return false
+	}
+
+	// ignore disabled built-in checks
+	if IsBuiltinNamespace(getModuleNamespace(module)) && s.disabledCheckIDs.Contains(metadata.ID) {
+		return false
+	}
+
+	if len(metadata.InputOptions.Selectors) == 0 && !metadata.Library {
+		s.logger.Warn(
+			"Module has no input selectors - it will be loaded for all inputs!",
```

Comment:
I recall reading somewhere on the Golang website about Go best practices for logging avoiding exclamation marks but I can't find it now. 

```suggestion
			"Module has no input selectors - it will be loaded for all inputs",
```

It also talked about case but my memory fails me as to what the recommendation was.

Reply:
Updated https://github.com/aquasecurity/trivy/pull/8478/commits/3ee747723753e94c6007dff56ca057be9f4353bd